test/e2e/clusterctl_upgrade.go

```diff
@@ -52,6 +54,8 @@ type ClusterctlUpgradeSpecInput struct {
 	BootstrapClusterProxy framework.ClusterProxy
 	ArtifactFolder        string
 	SkipCleanup           bool
+	PreUpgrade            func(managementClusterProxy framework.ClusterProxy)
+	PostUpgrade           func(managementClusterProxy framework.ClusterProxy)
```

Comment:
This allows for pre and post upgrade steps to be run by providers when running this spec.

For example, in cluster-api-provider-packet, we require users to execute both pre and post upgrade actions when moving from v1alpha3 to v1alpha4. This is to enable the use of a newer version of our cloud provider implementation, which needs to be done prior to the v1alpha4 upgrade, and an additional step to update the providerIDs on existing nodes after the v1alpha4 upgrade is finished.

Reply:
Thanks for this @detiber! (also cc @CecileRobertMichon which might be interested in this from CAPZ perspective)

> and an additional step to update the providerIDs on existing nodes after the v1alpha4 upgrade is finished

Curious about this one, do you have more details?

Reply:
@vincepri I don't think this applies to CAPZ,  but no objections from my side to enabling it for other providers as long as it's optional and doesn't break existing tests